src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -407,5 +406,176 @@ public static void SetLastWriteTime(string fullPath, DateTimeOffset time, bool a
 
         public static string[] GetLogicalDrives()
             => DriveInfoInternal.GetLogicalDrives();
+
+        internal static void CreateSymbolicLink(string path, string pathToTarget, bool isDirectory)
+        {
+            Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data = default;
+            FillAttributeInfo(pathToTarget, ref data, returnErrorOnNotFound: false);
+            if (data.dwFileAttributes != -1 &&
+                isDirectory != ((data.dwFileAttributes & Interop.Kernel32.FileAttributes.FILE_ATTRIBUTE_DIRECTORY) != 0))
+            {
+                throw new IOException(SR.IO_InconsistentLinkType);
+            }
```

Comment:
Yes we do. Windows does not enforce that the specified Target is a directory if `SYMBOLIC_LINK_FLAG_DIRECTORY` flag is passed. Perhaps we shouldn't enforce it as well but then someone could end up creating meaningless file symbolic links pointing to directories and vice versa. It's a weird scenario and I think it's more safe to throw rather than allow it. We can change it in the future if someone bring a reasonable enough scenario for it (it's not a hard breaking change).

Reply:
> Windows does not enforce that the specified Target is a directory if SYMBOLIC_LINK_FLAG_DIRECTORY flag is passed.

Wow! Interesting Windows design! What flag will the read operation return in this case - that it is a directory or a file? In other words if we use SYMBOLIC_LINK_FLAG_DIRECTORY but the target is a file, then a file link will be created and then read operation will return that it is a file link, right?

If this was not explicitly discussed during the API review I would rather prefer this behavior, because otherwise having this check and subsequent throw requires processing in user code (try-catch), in which case it is much better to do the check itself in user code (most likely it will already be there inevitably to exclude try-catch) and avoid throw - so we should remove unnecessary checks and throws and follow the behavior of the Windows API.

Moreover, it can be argued that a link is almost always created in a situation where the target is well known. I can't even think of a scenario where someone creates a link without knowing anything about the target.

Reply:
> _If this was not explicitly discussed during the API review_

We did mention that Windows differentiates between a symlink to a file and a symlink to a directory, hence why we needed to have separate creation APIs for directories and for files.

> _having this check and subsequent throw requires processing in user code (try-catch)_

We want to throw an exception because we think it's safer to tell the user early that they are using the wrong API for creating a symlink to the desired target.

> _it can be argued that a link is almost always created in a situation where the target is well known. I can't even think of a scenario where someone creates a link without knowing anything about the target._

I would consider this an argument in favor of throwing to let the user know they are creating a symlink using the wrong API, especially since they know the file was either a file or a directory.

Here are a couple of examples of the `mklink` command with the wrong target.

<details>
  <summary>Creating a directory link that points to a file</summary>

  ```
D:\repos\tests\symlinks>dir

 Directory of D:\repos\tests\symlinks

2021-06-22  09:09 PM    <DIR>          .
2021-06-22  09:09 PM    <DIR>          ..
2021-05-28  01:24 PM                 5 file.txt
2021-05-28  01:25 PM    <DIR>          folder
               1 File(s)              5 bytes
               3 Dir(s)  83,665,231,872 bytes free

D:\repos\tests\symlinks>mklink /d linkdir file.txt
symbolic link created for linkdir <<===>> file.txt
  ```
  Notice it says `<SYMLINKD>`:
  ```
D:\repos\tests\symlinks>dir

 Directory of D:\repos\tests\symlinks

2021-06-22  09:09 PM    <DIR>          .
2021-06-22  09:09 PM    <DIR>          ..
2021-05-28  01:24 PM                 5 file.txt
2021-05-28  01:25 PM    <DIR>          folder
2021-06-22  09:09 PM    <SYMLINKD>     linkdir [file.txt]
               1 File(s)              5 bytes
               4 Dir(s)  83,665,231,872 bytes free

D:\repos\tests\symlinks>cd linkdir
The directory name is invalid.
  ```
</details>

<details>
  <summary>Creating a file link that points to a directory</summary>

  ```
D:\repos\tests\symlinks>mklink linkfile folder
symbolic link created for linkfile <<===>> folder
  ```
  Notice it says `<SYMLINK>`:
  ```
D:\repos\tests\symlinks>dir

 Directory of D:\repos\tests\symlinks

2021-06-22  09:09 PM    <DIR>          .
2021-06-22  09:09 PM    <DIR>          ..
2021-05-28  01:24 PM                 5 file.txt
2021-05-28  01:25 PM    <DIR>          folder
2021-06-22  09:09 PM    <SYMLINKD>     linkdir [file.txt]
2021-06-22  09:09 PM    <SYMLINK>      linkfile [folder]
               2 File(s)              5 bytes
               4 Dir(s)  83,665,231,872 bytes free

D:\repos\tests\symlinks>cd linkfile
The directory name is invalid.
  ```
</details>

Reply:
> We want to throw an exception because we think it's safer to tell the user early that they are using the wrong API for creating a symlink to the desired target.

Why do you say about using _wrong_ API? It seems you consider **formal** scenario. Can you think **real** scenario where target is abstract (unknown type)? I can not. As result my conclusion is that the API should be universal and without throw since user/application always know that is the target. Ex.: an application unpack new version to version specific directory and then create a link from generic application folder to the new one.